### PR TITLE
add typehints for torchvision.datasets.folder

### PR DESCRIPTION
**Pull request by pmeier**

### CODE REVIEW

File: torchvision/datasets/folder.py [hunk 1]
```diff
@@ -31,16 +32,22 @@ def is_image_file(filename):
     return has_file_allowed_extension(filename, IMG_EXTENSIONS)
 
 
-def make_dataset(directory, class_to_idx, extensions=None, is_valid_file=None):
+def make_dataset(
+    directory: str,
+    class_to_idx: Dict[str, int],
+    extensions: Optional[Tuple[str, ...]] = None,
+    is_valid_file: Optional[Callable[[str], bool]] = None,
```
Comment (fmassa) [inline torchvision/datasets/folder.py:39]: Hum, from https://stackoverflow.com/questions/55231989/optional-union-in-type-hint this should not be possible?

Comment (fmassa) [reply torchvision/datasets/folder.py:39]: Ok, I missed one set of brackets while reading the code

**Comment by weikunhan** [comment]
Hello, 

I have created my own folder_sp. It works in a single GPU, but meets bugs in multiple GPUs. Anyone can tell me what is going on here?

_pickle.UnpicklingError: pickle data was truncated
  File "<string>", line 1, in <module>
  File "/usr/lib64/python3.6/multiprocessing/spawn.py", line 105, in spawn_main
    exitcode = _main(fd)
  File "/usr/lib64/python3.6/multiprocessing/spawn.py", line 115, in _main
    self = reduction.pickle.load(from_parent)
_pickle.UnpicklingError: pickle data was truncated
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/lib64/python3.6/multiprocessing/spawn.py", line 105, in spawn_main
    exitcode = _main(fd)
  File "/usr/lib64/python3.6/multiprocessing/spawn.py", line 115, in _main
    self = reduction.pickle.load(from_parent)
_pickle.UnpicklingError: pickle data was truncated
Traceback (most recent call last):
  File "main.py", line 102, in <module>
    main()
  File "main.py", line 45, in main
    classification.start(dataset_path, checkpoints_path, args, **CONFIG[args.dataset])
  File "/nfs/hpc/share/coe_hanweiku/xxxnet-pytorch/utils/classification.py", line 304, in start
    mp.spawn(main_worker, nprocs=ngpus_per_node, args=(ngpus_per_node, args))
  File "/nfs/hpc/share/coe_hanweiku/xxxnet-pytorch/venv2/lib64/python3.6/site-packages/torch/multiprocessing/spawn.py", line 200, in spawn
    return start_processes(fn, args, nprocs, join, daemon, start_method='spawn')
  File "/nfs/hpc/share/coe_hanweiku/xxxnet-pytorch/venv2/lib64/python3.6/site-packages/torch/multiprocessing/spawn.py", line 158, in start_processes
    while not context.join():
  File "/nfs/hpc/share/coe_hanweiku/xxxnet-pytorch/venv2/lib64/python3.6/site-packages/torch/multiprocessing/spawn.py", line 108, in join
    (error_index, name)
Exception: process 1 terminated with signal SIGKILL

My code is here: 
[folder_sp.txt](https://github.com/pytorch/vision/files/5072921/folder_sp.txt)

**Comment by pmeier** [comment]
Hi @weikunhan, 

as far as I can see the problem is not related to this PR in any way. Furthermore, the primary means for support is [our discussion forum](https://discuss.pytorch.org). To increase the chances that someone is able to help you, I suggest you compile a [minimal working example](https://en.wikipedia.org/wiki/Minimal_working_example).

**Comment by weikunhan** [comment]
@pmeier 

Thanks for helping~! I know what is reason for that!  I started a discussion in Pytorch forum: 

https://discuss.pytorch.org/t/torch-utils-data-dataloader-issue/92770

This problem was solved.

**Pull request merged**